### PR TITLE
Improve assignment to/from DefineMaps by avoiding copying special keys.

### PR DESCRIPTION
**Pull request by bmomberger-bitovi**
Covers #92 and other issues found when trying to convert between can.Map and DefineMap, which aren't currently playing nice with each other in certain cases.

These operations will help bridge that gap, and prevent some issues, but there are limitations to the current approach:
* I haven't touched on anything specific to DefineList yet, and I'm sure it has its own problems, but currently I'm having more, and more frequent, issues going between can.Map and DefineMap.

But here's what's been done so far:
* For cases where DefineMaps are being sent to other objects for assignment, the `constructor` property, the `"*"` defintition, and the prototype functions are made non-enumerable.  This prevents the prototype functions on target objects being overwritten, causing incorrect object behavior, and prevents a crash when trying to set the `constructor` property of can.Maps.
* For the reverse case, assigning other objects to DefineMaps, the special keys that matter to operating a DefineMap (_data, _bindEvents, constructor, and _cid) are scrubbed from input sources before assigning those inputs to a new or existing DefineMap.
* The `_bindings` special property was pushed into `__bindEvents` in can-event 3.3 so managing it is now not an issue.
* `define()` now defines its special properties (`__bindEvents`, `_cid`, and `constructor`) as not writable, to avoid other observable-like objects overwriting these properties incorrectly.
* The `.set(Object)` is accounted for, `new DefineMap(Object)` as well, and enough has been done to make `assign(DefineMap, Object)` work as intended, even if the second arg has special props.  However, testing could still be more comprehensive with another pair of eyes reviewing it.

### CODE REVIEW

**Comment by bmomberger-bitovi** [comment]
Also I converted indentation in all affected files to be tabs, which caused a lot of whitespace diff -- add `?w=1` to the URL when reviewing.

**Comment by bmomberger-bitovi** [comment]
So I think that perhaps instead of the special cases being done in this PR for the assign-from-other-object-to-DefineMap case, it would be better to define the special keys with object or unchanging values (_cid, _data, _bindEvents, and constructor) as non-writable in the property definition of each object instance during setup (this seems to already happen with _data but I'm not sure where), and use a special setter on the _bindings count to pass something that isn't a number when it should be incremented/decremented, avoiding copy operations.

Thoughts/comments?

File: define-helpers/define-helpers.js [hunk 1]
```diff
@@ -7,79 +7,87 @@ var canEvent = require("can-event");
...
+			});
+			return where;
+		} else {
+			return props;
+		}
+	},
+	removeSpecialKeys: function(map) {
+		if(map) {
+			["_data", "constructor", "_cid", "_bindings", "__bindEvents"].forEach(function(key) {
+				delete map[key];
```
Comment (justinbmeyer) [inline define-helpers/define-helpers.js]: I'm assuming `map` is cloned from elsewhere; otherwise we shouldn't be mutating an argument.

File: can-define.js [hunk 2]
```diff
@@ -64,7 +64,7 @@ module.exports = define = ns.define = function(objPrototype, defines, baseDefine
 	// for any value that has a default value.
 	replaceWith(objPrototype, "_data", function() {
 		var map = this;
-		var data = {};
+		var data = Object.create(null);
```
Comment (justinbmeyer) [inline can-define.js]: why is this needed?  (there's another issue where people are having problems with this technique #86)

File: define-helpers/define-helpers.js [hunk 3]
```diff
@@ -7,79 +7,87 @@ var canEvent = require("can-event");
...
+			props.each(function(value, prop){
+				where[prop] = value;
+			});
+			return where;
+		} else {
+			return props;
+		}
+	},
+	removeSpecialKeys: function(map) {
+		if(map) {
```
Comment (justinbmeyer) [inline define-helpers/define-helpers.js]: Is map an arg to `.set`? we shouldn't mutate arguments.

Comment (bmomberger-bitovi) [inline define-helpers/define-helpers.js]: see https://github.com/canjs/can-define/blob/improve-assignment/map/map.js#L32
https://github.com/canjs/can-define/blob/improve-assignment/map/map.js#L97
https://github.com/canjs/can-define/blob/improve-assignment/map/map.js#L125

The helper is called on the result of toObject and on a shallow clone of the map arg.  Because it had two different call cases, I opted to make the function not-smart and resolve that concern before calling it.

File: can-define.js (hunk 2 above)
Comment (bmomberger-bitovi) [inline can-define.js]: I disagree with the view of the person reporting that issue; in my estimation, accidental enumerable properties on the default prototype chain are more likely and more frustrating than not having default Object functions on reserved properties.  This instance above was meant to be boyscouting; I can remove it here, but if you took it to a consensus we'd likely opt for using this pattern for plain key/value map objects going forward.

Comment (bmomberger-bitovi) [inline can-define.js]: This line is reverted in latest.

**Comment by bmomberger-bitovi** [comment]
This PR now depends on https://github.com/canjs/can-event/pull/44 and must update the `can-event` reference in package.json once that PR is merged and a new version is deployed.

**Comment by justinbmeyer** [comment]
I'm worried that changing ._bindings to run through a getter / setter will cause performance issues.

File: define-helpers/define-helpers.js (hunk 1 above)
Comment (bmomberger-bitovi) [inline define-helpers/define-helpers.js]: ( https://github.com/canjs/can-define/pull/101#discussion_r93732186 covers this comment as well )

**Comment by bmomberger-bitovi** [comment]
> I'm worried that changing ._bindings to run through a getter / setter will cause performance issues.

I changed my approach and am now not using a setter here.  I think it's ready for re-review.  Do you have any more concerns?
